### PR TITLE
New rate_limit plugin for simple resource limitations

### PR DESCRIPTION
**Pull request by zwoop**
This current version has only one limiter, which implements
a basic active_in limitation. However, at least for now,
these connections that are queued wills still count against
the active connection metrics and limits that are in the core.

I'm open for a refactoring of this, if or when we want to have
different types of limiters. Similar to the policies for the
cache_promote plugin.

### CODE REVIEW

**Comment by zwoop** [comment]
[approve ci autest]

**Comment by zwoop** [comment]
Found a bug, TSContScheduleEveryOnPool() does not play nicely with config reloads and TSContDestroy() :-/.

File: plugins/experimental/rate_limit/limiter.h [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  bool
+  full() const
+  {
+    return (_size == max_queue);
+  }
+
+  void
+  push(TSHttpTxn txnp, TSCont cont)
+  {
+    QueueTime now = std::chrono::system_clock::now();
```
Comment (shukitchan) [inline plugins/experimental/rate_limit/limiter.h:104]: Should we use TShrtime() instead?

**Review by shukitchan — COMMENTED** [review]
I think we have proxy.config.net.max_connections_in / proxy.config.http.server.max_connections but they are applied globally. 

So we need this on a "per remap rules" basis ? Is this supposed to be a stop gap before we can make the above 2 "overridable" (i.e. can be changed by conf_remap) ?

File: plugins/experimental/rate_limit/limiter.h (hunk 1 above)
Comment (zwoop) [reply plugins/experimental/rate_limit/limiter.h:105]: I'd prefer not to :-). The APIs around TShrtime are lame :).

**Comment by zwoop** [comment]
[approve ci]

**Comment by zwoop** [comment]
[approve ci]

**Comment by zwoop** [comment]
[approve ci]

**Comment by zwoop** [comment]
Fixed the spelling errors, please review again :).

File: plugins/experimental/rate_limit/limiter.h [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+#include <atomic>
+#include <cstdio>
+#include <chrono>
+#include <cstring>
+#include <string>
+
+#include <ts/ts.h>
+
+constexpr char const PLUGIN_NAME[]  = "rate_limit";
+constexpr unsigned QUEUE_DELAY_TIME = 100; // Examine the queue every 100ms.
```
Comment (bneradt) [inline plugins/experimental/rate_limit/limiter.h:30]: Might want to consider:

```
constexpr auto QUEUE_DELAY_TIME = std::chrono::milliseconds{100};

...

_action = TSContScheduleEveryOnPool(_queue_cont, QUEUE_DELAY_TIME.count(), TS_THREAD_POOL_TASK);
```

**Review by bneradt — APPROVED** [review]
Looks great. I approve. Including a few minor suggestions for your consideration. Implement or drop them as you see fit.

File: doc/admin-guide/plugins/rate_limit.en.rst [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+The limit counters and queues are per remap rule only, i.e. there is
+(currently) no way to group transaction limits from different remap rules
+into a single rate limiter.
+
+All configuration is done via :file:`remap.config`, and the following options
+are available:
+
+.. program:: rate-limit
+
+.. option:: --limit
```
Comment (ywkaras) [inline doc/admin-guide/plugins/rate_limit.en.rst:38]: Why all the double dashes?

File: plugins/experimental/rate_limit/limiter.cc [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+RateLimiter::queue_process_cont(TSCont cont, TSEvent event, void *edata)
+{
+  RateLimiter *limiter = static_cast<RateLimiter *>(TSContDataGet(cont));
+  QueueTime now        = std::chrono::system_clock::now(); // Only do this once per "loop"
+
+  // Try to enable some queued txns (if any) if there are slots available
+  while (limiter->size() > 0 && limiter->reserve()) {
+    QueueItem item                  = limiter->pop();
+    TSHttpTxn txnp                  = std::get<0>(item);
+    std::chrono::microseconds delay = std::chrono::duration_cast<std::chrono::milliseconds>(now - std::get<2>(item));
```
[the remark this answers is not in the document]
Comment (zwoop) [reply plugins/experimental/rate_limit/limiter.cc:34]: Bryan hates Auto :-)

File: doc/admin-guide/plugins/rate_limit.en.rst (hunk 3 above)
Comment (zwoop) [reply doc/admin-guide/plugins/rate_limit.en.rst:38]: Cause that's how we do long-opts?

Comment (zwoop) [reply doc/admin-guide/plugins/rate_limit.en.rst:38]: Cause that's how we do this?

File: plugins/experimental/rate_limit/limiter.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (zwoop) [reply plugins/experimental/rate_limit/limiter.cc:34]: Bryan hates auto :).

File: doc/admin-guide/plugins/rate_limit.en.rst (hunk 3 above)
Comment (ywkaras) [reply doc/admin-guide/plugins/rate_limit.en.rst:38]: Do what?

Comment (ywkaras) [reply doc/admin-guide/plugins/rate_limit.en.rst:38]: Why not just omit them?  I could see the reason for double dashes before options for a global plugin but not so much for a remap plugin.  Seems redundant with the the @pparam.

Comment (zwoop) [reply doc/admin-guide/plugins/rate_limit.en.rst:38]: This is how like every plugin works. With the --'s, we can use standard libC getopt_long() APIs.

**Review by bneradt — APPROVED** [review]

File: plugins/experimental/rate_limit/limiter.h [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+#include <cstring>
+#include <string>
+
+#include <ts/ts.h>
+
+constexpr char const PLUGIN_NAME[] = "rate_limit";
+constexpr auto QUEUE_DELAY_TIME    = std::chrono::milliseconds{100}; // Examine the queue every 100ms
+
+using QueueTime = std::chrono::time_point<std::chrono::system_clock>;
+using QueueItem = std::tuple<TSHttpTxn, TSCont, QueueTime>;
```
Comment (ywkaras) [inline plugins/experimental/rate_limit/limiter.h:33]: Call me a loom smasher but I think I would just do:
```
struct QueueItem
{
  TSHttpTxn txnp;
  TSCont contp;
  QueueTime start_time;
};
```

**Review by bneradt — APPROVED** [review]

**Comment by zwoop** [comment]
Cherry-picked to v9.1.x branch.

**Comment by sudheerv** [comment]
Nice and useful plugin to have. 

A comment/observation though - 

1) Given that this is a remap plugin, it feels like the rate limiting of the transactions is effectively on the Origin transactions (at least, until H2 to Origin is supported)  as opposed to the client side. 
2) I also wonder if this is potentially close to what https://docs.trafficserver.apache.org/en/9.0.x/admin-guide/files/records.config.en.html#proxy-config-http-per-server-connection-max does already. There is also a queueing support with the outbound connection tracking.

Fwiw, we use a similar rate limiter plugin running in production with the difference being it is a global plugin and tracks concurrency against explicitly configured limits (ie not via remap.config). We've a bunch of business logic (including several sideways calls) that is run before remap state, so, queuing or limiting at remap is a little bit too late if we wanted to cap the resource usage more efficiently.

**Pull request merged**